Final Fantasy IV Free Enterprise: Add New Game

Rosalie-A opened the pull request:
## What is this fixing or adding?
Implementing a fork of the open source version of the Final Fantasy IV: Free Enterprise randomizer as a new game in Archipelago

## How was this tested?
Multiple private beta tests, some public ones, and single player games by both myself and contributing testers.

## If this makes graphical changes, please attach screenshots.
<img width="128" alt="AP 76989436201017225468 P1 Rosalie 2024-09-06 15 04 58" src="https://github.com/user-attachments/assets/4916615f-cd09-4ffc-b35b-65d53d6cca98">


I'm putting this up as a draft for now. I still need to implement the WebWorld side of things and make proper documentation, but I figured I'd put this up now to get eyes on it for anything that needs to be changed (especially re: how I'm doing locations and rules) while I do that so I'm not just spinning my wheels when I want to work on this.

ScipioWright commented:
It looks like you've added several files outside of your world, and have modified the Super Metroid as well. I would recommend fixing those issues

ScipioWright commented on worlds/ff4fe/__init__.py:
```diff
@@ -0,0 +1,397 @@
+from worlds.AutoWorld import World, WebWorld
+from BaseClasses import Region, Location, Entrance, Item, ItemClassification, MultiWorld, LocationProgressType
+from .rom import FF4FEProcedurePatch
+from ..generic.Rules import set_rule, add_rule, forbid_item, add_item_rule
```
```suggestion
from worlds.generic.Rules import set_rule, add_rule, forbid_item, add_item_rule
```
Need to use absolute imports for things outside of your apworld

ScipioWright commented on worlds/ff4fe/__init__.py:
```diff
@@ -0,0 +1,397 @@
+        self.rom_name_available_event = threading.Event()
+        self.chosen_character = "None"
+
+    def generate_early(self) -> None:
+        self.options.local_items.value.update(
+            ["Cecil", "Kain", "Rydia", "Tellah", "Edward", "Rosa",
+             "Yang", "Palom", "Porom", "Cid", "Edge", "Fusoya", "None"])
+        pass
```
Were you intending to do something else here too? If not, you should remove that pass at the end

ScipioWright commented on worlds/ff4fe/__init__.py:
```diff
@@ -0,0 +1,397 @@
+
+    def get_objective_count(self):
+        objective_count = 0
+        if self.options.ForgeTheCrystal.current_key == "true":
```
```suggestion
        if self.options.ForgeTheCrystal:
```
Or, you could do
`if self.options.ForgeTheCrystal is True:`

This is just a recommendation

ScipioWright commented on worlds/ff4fe/__init__.py:
```diff
@@ -0,0 +1,397 @@
+            if location.name not in locations.character_locations:
+                if location.name.startswith("Objective") or location.name == "Kokkol's House 2F -- Kokkol -- forge item":
+                    try:
+                        self.multiworld.get_location(location.name, self.player)
```
```suggestion
                        self.get_location(location.name)
```
There's a `world.get_location`, which is just a quicker version of what you were doing. Feel free to swap to it or not

ScipioWright commented on worlds/ff4fe/itempool.py:
```diff
@@ -0,0 +1,88 @@
+    return (result_pool, chosen_character)
+
+
+def create_character_pool(multiworld: World, chosen_character: str) -> List[str]:
```
```suggestion
def create_character_pool(world: World, chosen_character: str) -> List[str]:
```
okay I haven't seen this flipped the other way before. Would recommend doing world: World instead of multiworld: World

Rosalie-A replied:
and that's why it's a draft and I wanted eyes on it early because I knew I'd skim over half the stuff you mentioned at bare minimum. thank you, sincerely!

ScipioWright reviewed (commented):
Just a very brief skim, there's a LOT here

ScipioWright reviewed (commented):
Just a few more things

Silvris commented on .gitignore:
```diff
@@ -28,9 +29,12 @@
 *multisave
 *.archipelago
 *.apsave
-*.BIN
```
This is bad. There are ROM files currently supported by Archipelago that are *.bin. I would recommend just explicitly inverting the ignore to include whatever file is needed.

CaitSith2 commented on worlds/ff4fe/__init__.py:
```diff
@@ -191,6 +214,11 @@ def set_rules(self) -> None:
                               lambda item: item.name not in items.characters)
+                if self.options.ItemPlacement.current_key == "major_minor_split" and location.major_slot == False:
+                    add_item_rule(self.get_location(location.name),
+                                  lambda item: (item.classification & ItemClassification.progression == 0) and item.name != "DkMatter")
```
This line as is does not actually resolve the major-minor split and DkMatter in the item pool.

Instead, the code should be the following.
```python
add_item_rule(self.get_location(location.name),
                                  lambda item: (item.classification & ItemClassification.progression == 0) or item.name == "DkMatter")
```

CaitSith2 commented on worlds/ff4fe/__init__.py:
```diff
@@ -191,6 +214,11 @@ def set_rules(self) -> None:
+                if self.options.ItemPlacement.current_key == "major_minor_split" and location.major_slot == False:
+                    add_item_rule(self.get_location(location.name),
+                                  lambda item: (item.classification & ItemClassification.progression == 0) and item.name != "DkMatter")
+                else:
+                    print(location.name)
```
This seems to be debug printing.

CaitSith2 commented on worlds/ff4fe/__init__.py:
```diff
@@ -191,6 +214,11 @@ def set_rules(self) -> None:
                         continue
                 add_item_rule(self.get_location(location.name),
                               lambda item: item.name not in items.characters)
+                if self.options.ItemPlacement.current_key == "major_minor_split" and location.major_slot == False:
```
this can be rewritten to 
```python
if self.options.ItemPlacement.current_key == "major_minor_split" and not location.major_slot:
```

Also, a check to see if this location is in priority locations should be added here, since forbidding a priority location from having any progression items will cause a 100% generation failure rate.

Rosalie-A commented:
Alrighty! Above comments have been addressed, more testing has happened, features are now where they need to be, and I'm confident everything is where it needs to be, functionally. I throw this into you all's overworked and underpaid hands to scrutinize at your leisure.

ScipioWright commented on worlds/sm/data/sourceinfo.txt:
```diff
@@ -1,3 +0,0 @@
-SMBasepatch_prebuilt:
-- comes exactly from build/vanilla/ directory of https://github.com/lordlou/SMBasepatch
-- keep it in sync with the basepatch repo; do not modify the contents in this repo alone!
```
You should remove the edits to Super Metroid unless they're somehow required for FF4E

ScipioWright commented on worlds/ff4fe/FreeEnterpriseForAP/FreeEnt/f4c/lark/LICENSE:
```diff
@@ -0,0 +1,18 @@
+Copyright © 2017 Erez Shinan
```
lark is in here twice -- here and at `worlds/ff4fe/FreeEnterpriseForAP/f4c/lark/LICENSE`
It should probably only be in here once

Rosalie-A replied:
Turns out I didn't need the duplicate f4c module. anymore with me changing how the harp songs and z sprites work.

ScipioWright commented on worlds/ff4fe/__init__.py:
```diff
@@ -0,0 +1,488 @@
+
+        restricted_character_allow_locations = list(set(character_locations) - set(locations.restricted_character_locations)).copy()
+        restricted_character_forbid_locations = list(set(character_locations) - set(restricted_character_allow_locations)).copy()
+        self.multiworld.random.shuffle(restricted_character_allow_locations)
+        self.multiworld.random.shuffle(restricted_character_forbid_locations)
```
```suggestion
        self.random.shuffle(restricted_character_allow_locations)
        self.random.shuffle(restricted_character_forbid_locations)
```
You should use the world's random object rather than the multiworld's. This will avoid some nondeterministic behavior.

Rosalie-A replied:
Fixed.

ScipioWright commented on worlds/ff4fe/__init__.py:
```diff
@@ -0,0 +1,488 @@
+                        and self.options.ItemPlacement.current_key == "normal"
+                        and location.name not in locations.character_locations
+                        and not location.name.startswith("Objective")):
+                    add_rule(self.get_location(location.name),
+                             lambda state, tier=i: state.has_group("characters",
+                                                           self.player,
+                                                           FERules.logical_gating[tier]["characters"]) and
+                                           state.has_group("key_items",
+                                                           self.player,
+                                                           FERules.logical_gating[tier]["key_items"]))
```
Will it create issues if someone adds a character to their `start_inventory`? Since the randomizer might see "ah, you have 2 of character A and 1 of character B, so you can get this check that requires 3 characters".
If that would be an issue, you should use `has_group_unique` instead.

Rosalie-A replied:
Characters are not unique, so it's correct to not use has_group_unique. While it's true adding them to start inventory will cause the generator to think there are checks that are now "in logic", this will have little effect in practice: 

- characters in start inventory don't do anything (they're items for convenience's sake on the part of options and the code, not because they're equivalent to a Life potion or whatnot), 
- the gating logic is about spreading things out so at least some progression is located in less difficult areas.
- that gating logic still requires key items as well, so nothing will change.
- and even if there were changes, it makes things slightly more difficult for the player but doesn't render the game unbeatable.

ScipioWright commented on worlds/ff4fe/__init__.py:
```diff
@@ -0,0 +1,488 @@
+                    "location_data": location_data.to_json(),
+                    "item_data": ItemData.create_ap_item().to_json(),
+                    "item_name": location.item.name,
+                    "player_name": self.multiworld.player_name[location.item.player]
```
```suggestion
                    "player_name": self.player_name,
```

Rosalie-A replied:
The placement dictionary includes names for all player's items, not just the FE player. I need to use the lookup to get those names too so text like "got Silver Sword for Player5" can be written into the game.

Silvris commented on worlds/ff4fe/topology.py:
```diff
@@ -0,0 +1,77 @@
+    "LunarPalace"
+]
+
+overworld_areas = [area for area in areas if area not in [*hook_areas, *underworld_areas, *moon_areas]]
```
```suggestion
overworld_areas = [area for area in areas if area not in [*hook_areas, *underworld_areas, *moon_areas]]

```

Silvris reviewed (commented):
Didn't make it through everything, but I'm out of time for review today.

Rosalie-A commented on worlds/ff4fe/__init__.py:
```diff
@@ -0,0 +1,488 @@
+    settings: typing.ClassVar[FF4FESettings]
+
+    web = FF4FEWebWorld()
+    topology_present = True
```
I definitely maybe thought I had a good reason at one point, but mostly it let me confirm I set the regions and locations up correctly, so it can be removed, yeah. Will do so.

Rosalie-A commented:
Updated according to PR comments. I could click "resolve" on all of them but not sure if that's desired.

I also removed the FEExternalDataLoader program in favor of ~~the way I should have done it before~~ a better way to implement external harp songs and z sprites without packaging derivative material with the project.

Rosalie-A commented:
alrighty, in the interest of making life easier for the reviewers, I embarked on a not-paticularly-large project but it still took a hot minute.

Free Enterprise is now separated out from the APWorld and is fetched separately by requirements.txt. This should make the PR a lot more reasonable for y'all since you don't have a giant-ass randomizer to look over. There's very little change (a bit in the patching in rom.py, some data files bundled in for generation instead of being pulled from FE) compared to previous commits otherwise.

Rosalie-A commented:
Updated with changes and improvements based on feedback during testing. Barring some unusual new showstopper bug appearing, this will be the final update to this PR (aside from feedback here, of course) as future improvements will be saved for when the upstream rando updates and I need to support that, and that'll probably be better suited for a future PR.
